nncf/experimental/common/pruning/operations.py

```diff
@@ -0,0 +1,700 @@
+                         tensor_processor: Type[NNCFPruningBaseTensorProcessor]) -> None:
+        input_masks = get_input_masks(node, graph)
+        for m in input_masks:
+            if m:
+                m.invalidate_groups()
+        node.data['output_mask'] = None
+
+class ExpandAsPruningOp(BasePruningOp):
+    @classmethod
+    def mask_propagation(cls, node: NNCFNode, graph: NNCFGraph,
+                         tensor_processor: Type[NNCFPruningBaseTensorProcessor]) -> None:
+        input_edges = graph.get_input_edges(node)
+        assert len(input_edges) == 2, 'expand should always have 2 inputs'
+        input_to_expand = input_edges[0].from_node
+        mask = input_to_expand.data.get('output_mask')
+        propagated_mask = None
+        if mask is not None:
+            nncf_logger.warning(f"expand_as is applied to the node with propagation mask. Currently, it's not supported "
+                                "and mask is invalidated. node_name={node.node_name}")
+            mask.invalidate_groups()
+        input_to_get_shape = input_edges[1].from_node
+        mask: PropagationMask = input_to_get_shape.data.get('output_mask')
+        if mask is not None:
+            target_shape = input_edges[1].tensor_shape
+            source_shape = input_edges[0].tensor_shape
+            for dim, groups in mask.dim_groups_map.items():
+                if target_shape[dim] == source_shape[dim]:
+                    nncf_logger.warning(f"expand_as takes the shape from the node with propagation mask and pruning "
+                                        "dimension in the mask matches the dimension in the expanded input. Currently, "
+                                        "it's not supported and mask is invalidated. node_name={node.node_name}")
+                    for group in groups:
+                        group.invalidate()
+            propagated_mask = mask
+        node.data['output_mask'] = propagated_mask
+
+
+class ScatterPruningOp(BasePruningOp):
+    @classmethod
+    def mask_propagation(cls, node: NNCFNode, graph: NNCFGraph,
+                         tensor_processor: Type[NNCFPruningBaseTensorProcessor]) -> None:
+        input_masks = [input_edge.from_node.data.get('output_mask') for input_edge in graph.get_input_edges(node)]
+        assert len(input_masks) == 2, 'expect that masked_fill should always have 2 inputs'
+        i1, i2 = input_masks
+        propagated_mask = None
+        if i1 != i2:
+            nncf_logger.warning(f"expand_as takes the shape from the node with propagation mask and pruning "
+                                "dimension in the mask matches the dimension in the expanded input. Currently, "
+                                "it's not supported and mask is invalidated. node_name={node.node_name}")
+            if i1 != None:
+                i1.invalidate_groups()
+        else:
+            propagated_mask = i1
+        node.data['output_mask'] = propagated_mask
```

Comment:
![image](https://user-images.githubusercontent.com/4014476/229458075-4fc2e754-7304-4170-b52f-4c2ce68e8a17.png)
limited masked fill for the case when on both branches the same mask for the sake of simple implementation, since it's not often presented in the models.

Reply:
Important: Ngraph pruning transformation most likely won't work in such cases, masks will be invalidated. Should be checked

Reply:
Discussed in person, currently assume that mask is always zero and doesn't affecrt on inference